File: app/src/main/java/com/digis/android/task/ChartType.java
package com.digis.android.task;

import java.util.Random;

import androidx.annotation.DrawableRes;

public enum ChartType {
    RSRP(-140f, -60f, 0.3f) {
        @Override
        int getProgressDrawable(int value) {
            if (value >= -80)
                return EXCELLENT_SIGNAL;
            else if (value >= -90)
                return GOOD_SIGNAL;
            else if (value > -100)
                return FAIR_SIGNAL;
            else
                return POOR_SIGNAL;
        }
    },
    RSRQ(-30f, 0f, 0.5f) {
        @Override
        int getProgressDrawable(int value) {
            if (value >= -10)
                return EXCELLENT_SIGNAL;
            else if (value >= -15)
                return GOOD_SIGNAL;
            else if (value > -20)
                return FAIR_SIGNAL;
            else
                return POOR_SIGNAL;
        }
    },
    SINR(-10f, 30f, 0.8f) {
        @Override
        int getProgressDrawable(int value) {
            if (value >= 20)
                return EXCELLENT_SIGNAL;
            else if (value >= 13)
                return GOOD_SIGNAL;
            else if (value > 0)
                return FAIR_SIGNAL;
            else
                return POOR_SIGNAL;
        }
    };

    private static final Random RANDOM = new Random();

    @DrawableRes
    private static final int EXCELLENT_SIGNAL = R.drawable.green_gradient_progress;

    @DrawableRes
    private static final int GOOD_SIGNAL = R.drawable.yellow_gradient_progress;

    @DrawableRes
    private static final int FAIR_SIGNAL = R.drawable.orange_gradient_progress;

    @DrawableRes
    private static final int POOR_SIGNAL = R.drawable.red_gradient_progress;

    private final float min;
    private final float max;
    private final float scalar;

    ChartType(float min, float max, float scalar) {
        this.min = min;
        this.max = max;
        this.scalar = scalar;
    }

    public float getMin() {
        return min;
    }

    public float getMax() {
        return max;
    }

    public int scale(int value) {
        float random;
        float fraction = RANDOM.nextFloat() * scalar * value;
        if (RANDOM.nextBoolean())
            random = value + fraction;
        else
            random = value - fraction;
        // Ensuring the random value is between the specified range
        if (random < min)
            random = min;
        else if (random > max)
            random = max;

        return (int)random;
    }

    @DrawableRes
    abstract int getProgressDrawable(int value);
}

File: app/src/main/java/com/example/android/rtdp/implementation/kotlin/ChartType.kt
package com.example.android.rtdp.implementation.kotlin

import androidx.annotation.DrawableRes
import com.example.android.rtdp.R
import java.util.*

private val RANDOM: Random = Random()

@DrawableRes
private const val EXCELLENT_SIGNAL = R.drawable.green_gradient_progress

@DrawableRes
private const val GOOD_SIGNAL = R.drawable.yellow_gradient_progress

@DrawableRes
private const val FAIR_SIGNAL = R.drawable.orange_gradient_progress

@DrawableRes
private const val POOR_SIGNAL = R.drawable.red_gradient_progress

enum class ChartType(val min: Float, val max: Float, private val scalar: Float) {
    RSRP(-140f, -60f, 0.3f) {
        @DrawableRes
        override fun getProgressDrawable(value: Int): Int = when {
            value >= -80 -> EXCELLENT_SIGNAL
            value >= -90 -> GOOD_SIGNAL
            value > -100 -> FAIR_SIGNAL
            else -> POOR_SIGNAL
        }
    },
    RSRQ(-30f, 0f, 0.5f) {
        @DrawableRes
        override fun getProgressDrawable(value: Int): Int = when {
            value >= -10 -> EXCELLENT_SIGNAL
            value >= -15 -> GOOD_SIGNAL
            value > -20 -> FAIR_SIGNAL
            else -> POOR_SIGNAL
        }

    },
    SINR(-10f, 30f, 0.8f) {
        @DrawableRes
        override fun getProgressDrawable(value: Int): Int = when {
            value >= 20 -> EXCELLENT_SIGNAL
            value >= 13 -> GOOD_SIGNAL
            value > 0 -> FAIR_SIGNAL
            else -> POOR_SIGNAL
        }
    };

    fun scale(value: Int): Int {
        val fraction = RANDOM.nextFloat() * scalar * value
        val random = if (RANDOM.nextBoolean()) value + fraction else value - fraction
        // Ensuring the random value is between the specified range
        return random.coerceIn(min, max).toInt()
    }

    @DrawableRes
    abstract fun getProgressDrawable(value: Int): Int
}